js/web/lib/wasm/jsep/webgpu/ops/binary-like-util.ts

```diff
@@ -0,0 +1,120 @@
+import {GpuDataType, ProgramInfo, ProgramInfoLoader, ProgramMetadata} from '../types';
+
+import {ShaderHelper} from './common';
+
+type BuiltinFunctionName = string;
+export type BinaryCustomExpression = (expressionA: string, expressionB: string, expressionC?: string) => string;
+export type BinaryFunctionCall = BuiltinFunctionName|BinaryCustomExpression|{
+  scalar: BinaryCustomExpression;
+  vector: BinaryCustomExpression;
+};
+
+type CreateOpProgramShader =
+    (shaderHelper: ShaderHelper, inputs: readonly TensorView[], dimsOutput: readonly number[], vectorize: boolean,
+     doBroadcast: boolean, funcCall: BinaryFunctionCall, typeOutput: number, additionalImplementation?: string) =>
+        string;
+
+/* eslint-disable no-param-reassign */
+const createOpProgramInfo =
+    (metadata: ProgramMetadata, inputs: readonly TensorView[], funcCall: BinaryFunctionCall,
+     createOpProgramShader: CreateOpProgramShader, additionalImplementation?: string,
+     outputDataType?: number): ProgramInfo => {
+      const a = inputs.length === 3 ? inputs[1] : inputs[0];
+      const b = inputs.length === 3 ? inputs[2] : inputs[1];
+      if (outputDataType == null) {
+        outputDataType = inputs.length === 3 ? inputs[1].dataType : inputs[0].dataType;
+      }
+
+      const isBroadcast = !ShapeUtil.areEqual(a.dims, b.dims);
+      let outputShape = a.dims;
+      let outputSize = ShapeUtil.size(a.dims);
+
+      let vectorize = false;
+
+      // TODO: deal with zero-sized tensors (eg. dims=[1,0])
+      if (isBroadcast) {
+        const calculatedShape = BroadcastUtil.calcShape(a.dims, b.dims, false);
+        if (!calculatedShape) {
+          throw new Error('Can\'t perform binary op on the given tensors');
+        }
+        outputShape = calculatedShape;
+        outputSize = ShapeUtil.size(outputShape);
+
+        // check whether vectorize can be enabled
+        let sharedDimension = 1;
+        for (let i = 1; i < outputShape.length; i++) {
+          const dimA = a.dims[a.dims.length - i] ?? 1;
+          const dimB = b.dims[b.dims.length - i] ?? 1;
+          if (dimA === dimB) {
+            sharedDimension *= dimA;
+          } else {
+            break;
+          }
+        }
+        if (sharedDimension % 4 === 0) {
+          vectorize = true;
+        }
+      } else {
+        // element-wise
+        vectorize = true;
+      }
+
+      return {
+        ...metadata,
+        getShaderSource: (shaderHelper) => createOpProgramShader(
+            shaderHelper, inputs, outputShape, vectorize, isBroadcast, funcCall, outputDataType as number,
+            additionalImplementation),
+        outputs: [{dims: outputShape, dataType: outputDataType, gpuDataType: GpuDataType.default}],
+        dispatchGroup: () =>
+            ({x: Math.ceil(outputSize / 64 /* workgroup size */ / (vectorize ? 4 : 1) /* vec size */)})
+      };
+    };
+
+// This is used for ops like binary, where.
+export const createOpProgramInfoLoader =
```

Comment:
maybe `createBinaryLikeProgramInfoLoader` is better? ( a little bit too long but it's fine )

Reply:
done